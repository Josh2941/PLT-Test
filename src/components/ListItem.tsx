import React from 'react'
import { Button, Image, StyleSheet, Text, View } from "react-native"

export type ListItemProps = {
    id: number
    colour: string
    name: string
    price: number
    img: string
    selectedCount?: number
}

type ItemProps = ListItemProps & {
    onPressIncrement?: (item: ListItemProps) => void
    onPressDecrement?: (item: ListItemProps) => void
}

export const ListItem = (props: ItemProps) => {
    const { id, name, colour, price, img, selectedCount, onPressIncrement, onPressDecrement } = props;
    return (
        <View style={styles.container} key={id}>
            <Image style={styles.image} source={{ uri: img }} /> 
            <View style={{ flex: 1 }}>
                <Text testID='listItemName' style={styles.name}>{name}</Text>
                <View style={styles.bottomRow}>
                    <View>
                        <Text testID='listItemColor' style={{ marginBottom: 10 }}>{colour}</Text>
                        <Text testID='listItemPrice' style={{ marginBottom: 10 }}>{price}</Text>
                    </View>
                    <View style={{ flexDirection: 'row', alignItems: 'center' }}>
                        {selectedCount && selectedCount > 0 ?
                        <>
                            <Button testID='listItemDecermentButton' title="-" onPress={() => onPressDecrement?.(props as ListItemProps)}/>
                            <Text testID='listItemCount'>{selectedCount}</Text>
                            <Button testID='listItemIncermentButton' title="+" onPress={() => onPressIncrement?.(props as ListItemProps)}/>
                        </>
                        : 
                        <Button testID='listItemAddToCartButton' title="Add To Cart" onPress={() => onPressIncrement?.(props as ListItemProps)}/>}
                    </View>
                </View>
            </View>
        </View>
    )
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: 'row',
        padding: 10
    },
    image: {
        width: 100,
        height: 100
    },
    name: {
        width: '80%',
        marginBottom: 10
    },
    bottomRow: {
        flexDirection: 'row',
        justifyContent: 'space-between'
    }
})