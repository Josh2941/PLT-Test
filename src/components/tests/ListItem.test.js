import renderer, { act } from 'react-test-renderer'
import { ListItem } from '../ListItem';

const mockOnPressIncrement = jest.fn();
const mockOnPressDecrement = jest.fn();

const MOCK_PROPS = {
    id: 1,
    colour: "Black",
    name: "Black Sheet Strappy Textured Glitter Bodycon Dress",
    price: 10,
    img: "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
    onPressIncrement: mockOnPressIncrement,
    onPressDecrement: mockOnPressDecrement,
}

describe('Test ListItem', () => {
    test('renders correctly', () => {
        const tree = renderer.create(<ListItem {...MOCK_PROPS} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    test('show list item name correctly', () => {
        const component = renderer.create(<ListItem {...MOCK_PROPS} />).root;
        const listItemName = component.findByProps({ testID: 'listItemName' });
        expect(listItemName.props.children).toBe(MOCK_PROPS.name);
    });
    test('show list item colour correctly', () => {
        const component = renderer.create(<ListItem {...MOCK_PROPS} />).root;
        const listItemColor = component.findByProps({ testID: 'listItemColor' });
        expect(listItemColor.props.children).toBe(MOCK_PROPS.colour);
    });
    test('show list item price correctly', () => {
        const component = renderer.create(<ListItem {...MOCK_PROPS} />).root;
        const listItemPrice = component.findByProps({ testID: 'listItemPrice' });
        expect(listItemPrice.props.children).toBe(MOCK_PROPS.price);
    });
    test('to perform decrement action correctly', () => {
        const newMockProps = {
            ...MOCK_PROPS,
            selectedCount: 2,
        };
        const component = renderer.create(<ListItem {...newMockProps} />).root;
        const listItemDecermentButton = component.findByProps({ testID: 'listItemDecermentButton' });
        act(() => {
            listItemDecermentButton.props.onPress();
        })
        expect(mockOnPressDecrement).toHaveBeenCalled();
    });
    test('to perform increment action correctly', () => {
        const newMockProps = {
            ...MOCK_PROPS,
            selectedCount: 2,
        };
        const component = renderer.create(<ListItem {...newMockProps} />).root;
        const listItemIncermentButton = component.findByProps({ testID: 'listItemIncermentButton' });
        act(() => {
            listItemIncermentButton.props.onPress();
        })
        expect(mockOnPressIncrement).toHaveBeenCalled();
    });
    test('to perform add to cart action correctly', () => {
        const component = renderer.create(<ListItem {...MOCK_PROPS} />).root;
        const listItemAddToCartButton = component.findByProps({ testID: 'listItemAddToCartButton' });
        act(() => {
            listItemAddToCartButton.props.onPress();
        })
        expect(mockOnPressIncrement).toHaveBeenCalled();
    });
})
