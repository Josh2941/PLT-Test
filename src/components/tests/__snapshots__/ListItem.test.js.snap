// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ListItem renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
      "flexDirection": "row",
      "padding": 10,
    }
  }
>
  <Image
    source={
      {
        "uri": "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
      }
    }
    style={
      {
        "height": 100,
        "width": 100,
      }
    }
  />
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <Text
      style={
        {
          "marginBottom": 10,
          "width": "80%",
        }
      }
      testID="listItemName"
    >
      Black Sheet Strappy Textured Glitter Bodycon Dress
    </Text>
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View>
        <Text
          style={
            {
              "marginBottom": 10,
            }
          }
          testID="listItemColor"
        >
          Black
        </Text>
        <Text
          style={
            {
              "marginBottom": 10,
            }
          }
          testID="listItemPrice"
        >
          10
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="listItemAddToCartButton"
        >
          <View
            style={
              [
                {},
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
              Add To Cart
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
